l1-contracts/deploy-scripts/AddressIntrospector.sol

```diff
@@ -0,0 +1,231 @@
+        info.ethTokenAssetId = _assetRouter.ETH_TOKEN_ASSET_ID();
+    }
+
+    function getBaseTokenRoute(
+        IBridgehub _bridgehub,
+        uint256 _chainId
+    ) public view returns (BaseTokenRoute memory info) {
+        info.baseTokenAssetId = _bridgehub.baseTokenAssetId(_chainId);
+        address ar = _bridgehub.assetRouter();
+        info.assetHandlerAddress = IAssetRouterBase(ar).assetHandlerAddress(info.baseTokenAssetId);
+        if (info.assetHandlerAddress != address(0)) {
+            info.baseTokenAddress = IL1BaseTokenAssetHandler(info.assetHandlerAddress).tokenAddress(
+                info.baseTokenAssetId
+            );
+        }
+    }
+
+    function getZkChainFacetAddresses(IZKChain _zkChain) public view returns (address[] memory) {
+        return _zkChain.facetAddresses();
+    }
+
+    function getL1ERC20BridgeAddresses(
+        IL1ERC20Bridge _bridge
+    ) public view returns (L1ERC20BridgeAddresses memory info) {
+        info.l1Nullifier = address(_bridge.L1_NULLIFIER());
+        info.l1AssetRouter = address(_bridge.L1_ASSET_ROUTER());
+        info.l1NativeTokenVault = address(_bridge.L1_NATIVE_TOKEN_VAULT());
+        info.l2TokenBeacon = _bridge.l2TokenBeacon();
+        info.l2Bridge = _bridge.l2Bridge();
+        info.eraChainId = _tryUint256(address(_bridge), "ERA_CHAIN_ID()");
+        info.l2TokenProxyBytecodeHash = _tryBytes32(address(_bridge), "l2TokenProxyBytecodeHash()");
+    }
+
+    /// @notice Convenience method to fetch everything for a specific chainId via a Bridgehub instance
+    function getAllForChain(
+        IBridgehub _bridgehub,
+        uint256 _chainId
+    )
+        external
+        view
+        returns (
+            BridgehubAddresses memory bh,
+            CTMAddresses memory ctm,
+            ZkChainAddresses memory zk,
+            AssetRouterAddresses memory ar,
+            BaseTokenRoute memory baseRoute,
+            address[] memory zkFacets,
+            L1ERC20BridgeAddresses memory legacyBridge
+        )
+    {
+        bh = getBridgehubAddresses(_bridgehub);
+
+        address ctmAddr = _bridgehub.chainTypeManager(_chainId);
+        ctm = getCTMAddresses(IChainTypeManager(ctmAddr));
+
+        address zkAddr = _bridgehub.getZKChain(_chainId);
+        zk = getZkChainAddresses(IZKChain(zkAddr));
+
+        ar = getAssetRouterAddresses(IL1AssetRouter(payable(_bridgehub.assetRouter())));
+        baseRoute = getBaseTokenRoute(_bridgehub, _chainId);
+        zkFacets = getZkChainFacetAddresses(IZKChain(zkAddr));
+
+        // Optional: if legacy ERC20 bridge is known/set on the asset router, caller can provide it separately.
+    }
+
+    function _getBridgehubAdmin(IBridgehub _bridgehub) private view returns (address) {
+        // Bridgehub exposes `admin()` as public variable; access via staticcall interface-less
+        (bool ok, bytes memory data) = address(_bridgehub).staticcall(abi.encodeWithSignature("admin()"));
```

Comment:
I really think these static calls are bad. We should do it based on protocol version for most cases.